src/v/kafka/server/handlers/produce.cc

```diff
@@ -234,6 +234,77 @@ ss::future<produce_response::partition> finalize_request_with_error_code(
         .partition_index = ntp.tp.partition, .error_code = ec});
 }
 
+/**
+ * @brief Validate the timestamps of the batch, they have to be within a window
+ * to the broker's time. returns the new timestamp to set as max_timestamp to
+ * the batch, if present
+ */
+static auto validate_batch_timestamps(
+  model::ntp const& ntp,
+  model::record_batch const& batch,
+  model::timestamp_type timestamp_type,
+  net::server_probe& probe) -> std::optional<model::timestamp> {
+    // we compute in std::chrono::timepoints, we print in model::timestamps
+    auto broker_time = model::timestamp::now();
+    auto broker_timepoint = model::duration_since_epoch(broker_time);
+
+    // alert if first_timestamp is too far in the past
+    // default value for threshold basically disables this check, so it's
+    // wrapped in a if check
+    if (auto max_before = config::shard_local_cfg()
+                            .log_message_timestamp_alert_before_ms.value();
+        unlikely(max_before < std::chrono::milliseconds::max())) {
+        auto first_timestamp = batch.header().first_timestamp;
+        auto first_timepoint = model::duration_since_epoch(first_timestamp);
+        if (
+          broker_timepoint > first_timepoint
+          && std::chrono::duration_cast<std::chrono::milliseconds>(
+               broker_timepoint - first_timepoint)
+               > max_before) {
+            // generate an alert
+            klog.warn(
+              "{} too far in the past. broker time: {} timestamp: {}",
```

Comment:
nit: `produce request timestamp for {} was past the alert threshold, broker time: {}, timestamp: {}`